### ISSUE TITLE
Plugin interoperability issues (a.k.a. global event queue design?)

### ISSUE THREAD

**Issue by gedankenexperimenter** [keyboardio/Kaleidoscope#919]
## Background

We've had a few discussions about the idea of some sort of global event queue in the past, but so far nothing much has come of it. While looking into issues with TapDance (e.g. #806), I decided to have a go at redesigning it to use a Qukeys-style `KeyAddrEventQueue`, which would suit the purpose well. This brought to mind the somewhat vague discussions about a "global event queue" we've had, and got me thinking more specifically about what that would mean, and what mechanisms it would need to have in order to facilitate better plugin interactions. I'm opening this issue as a place to have a design discussion on the topic, in hopes that we can collaborate on something that will work better than what we've got now.

## Existing problems

I'd like to start by identifying problems that the current system has. There are many plugins that generate, alter, consume, or delay and restart key events, all of which have the potential to cause problems, not only for other plugins, but even for the plugin that called `handleKeyswitchEvent()` itself.

In addition, there's the larger problem of plugin interactions being sensitive to initialization order. For example, if a pair of plugins each changes the key value, the first one to get its event handler called can send information to the second one, but if the later one changes the key value to something handled specially by the first one, it will fail. Also, a plugin might assume that an event that it let pass through will have a specific effect on the final output to the host, but a later plugin could violate that assumption by returning `EVENT_CONSUMED`.

### Current strategies

Each plugin that generates key events (i.e. calling `handleKeyswitchEvent()`) employs some strategy for not acting on its own generated events. In most cases, that means setting the `INJECTED` bit in the keyswitch state parameter, then ignoring any event with that bit set in its `onKeyswitchEvent()` handler. This works well for a single plugin in isolation, but it's no good if, for example, the TapDance plugin generates an event with a OneShot key. TapDance would generate and `INJECTED` event, and OneShot would ignore that event because it also generates events, and uses the same strategy to avoid re-processing the same event.

Qukeys uses a different strategy, because I wanted it to be able to send events that would not bypass that very check, and instead get processed by other plugins. Instead of setting the `INJECTED` state bit, it sets an internal flag, calls `handleKeyswitchEvent()`, then unsets that flag, and ignores events if that flag is set. Unfortunately, this doesn't really do the trick either. The problem is that if TapDance was rewritten like Qukeys, both plugins would delay events and release them later, after that flag was unset, and could then get into a loop, forever queuing, delaying, and recreating the same event, passing it back and forth. It's essentially the same problem as `INJECTED`, but with different failure modes (plugins end up doing too much, rather than not enough).

## Ideas from Kaleidoglyph

Some of you may recall that, in an effort to better understand Kaleidoscope, I embarked on the ridiculous task of rewriting it to suit my own preferences. In doing so, I thought a lot about these issues, and came up with some solutions. I don't have anything that's perfect; I'm pretty sure that it's not possible to have a perfect system if third-party plugins are considered, but I did come up with some ideas that might prove useful to at least mitigate these issues in Kaleidoscope:

### Plugin event handler IDs

I set things up so that each plugin gets a unique identifier. Whenever a plugin emits a key event, it identifies itself, so the controller can avoid calling that plugin's event handler in response. The plugins therefore don't have to take responsibility for this task themselves.

### Two separate event handler hooks

Instead of having just one hook function for handling key events, I separated plugins into two categories: physical and logical. Different guarantees are promised by handlers in the two categories. 

"Physical" keyswitch event handlers can delay events, suppress events, change the order of events (in some cases), and change `Key` values, but they promise not to create new events that don't correspond to physical keyswitch press or release events. The order of these plugins is important, because any time one of them releases a queued event, that event gets handled first by the next plugin in the order; the calling plugin and all preceding plugins don't get their event handlers called, ensuring that each "physical" key event handler sees a given event only once. (Examples: Qukeys, TapDance)

"Logical" key event handlers can generate new events, but these events don't get handled by the "physical" event handlers at all. They are permitted to generate new events (that don't correspond one-to-one with physical keyswitch events), but they don't queue or delay events. They can change key values, and abort events, but if not aborted, a report will get triggered based on the resulting value. I set things up so that the controller checks the key value after each such handler is called, and restarts processing of these handlers if that value changes. It then skips any handler (using a bitmask) that has already processed the event, to prevent loops. This makes the plugin order much, much less important here. (Examples: OneShot, ShapeShifter, Macros)

## Other ideas

I know that implementing a Kaleidoglyph-style event handler, where much more of the logic is handled centrally, might be very difficult to introduce to Kaleidoscope, with its TMP plugin system, but there might be some other tricks we could use to at least mitigate the problem.

### Event IDs

Maybe we could assign each toggle-on or toggle-off event a sequential ID number. Plugins that call `handleKeyswitchEvent()` could then check that ID when processing an event and use it to help decide whether or not to process an event. (Perhaps it could reject events in the "recent past", where each plugin could track whatever the "most recent" event it had processed). The trouble is that this might only work for events that correspond to physical keyswitch toggles; it would be easy to get out of sync if Macros sends a long series of events, for example.

I think it might work to combine event IDs with the physical/logical separation I came up with for Kaleidoglyph. Plugins using the physical event handlers would be using ID numbers guaranteed to correspond to physical events, which would then only have problems if those events get their order changed. Even that could be solved by the plugins also swapping the event IDs.

### Bone fide event queue?

None of this is really what I would describe as a "global event queue", but maybe one would help. It could be implemented as a ring buffer, and each plugin could keep track of an index of where it is in the buffer, but things get real hairy if we include "injected" events in that queue, or even swap the order of two of the events (like I just submitted a PR for to fix a Qukeys bug).

---

[One more note re: Kaleidoglyph, but slightly outside the scope of the problem above: it doesn't store keyboard state in a HID report, but instead keeps it in an "active keys" array, functionally the same as my recent proposal for altering the keymap cache. And it only sends reports in response to key state changes (or timers), so if multiple events happen in the same scan cycle, it would send multiple reports in that scan cycle, with each plugin's event handler(s) being called once per event, rather than once per active key. This greatly simplifies the logic required when programming. There's no global event _queue_, but the whole thing is still event _driven_.]

**Comment by gedankenexperimenter** [comment]
Most of these ideas have now been merged into Kaleidoscope.  We still might want to implement a global event queue, but that's likely to be brought up as a new PR regardless of this issue being open, so I'm closing this one now.
